## Preprocessing

In [14]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [15]:
# Drop the non-beneficial ID columns, 'EIN'.

application_df = application_df.drop(columns = ['EIN'], axis = 1)

In [16]:
# Determine the number of unique values in each column.

[(i, len(application_df[i].unique())) for i in application_df.columns]

[('NAME', 19568),
 ('APPLICATION_TYPE', 17),
 ('AFFILIATION', 6),
 ('CLASSIFICATION', 71),
 ('USE_CASE', 5),
 ('ORGANIZATION', 4),
 ('STATUS', 2),
 ('INCOME_AMT', 9),
 ('SPECIAL_CONSIDERATIONS', 2),
 ('ASK_AMT', 8747),
 ('IS_SUCCESSFUL', 2)]

In [17]:
# Look at APPLICATION_TYPE value counts for binning

application_df.APPLICATION_TYPE.value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [18]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(application_df.APPLICATION_TYPE.value_counts()[application_df.APPLICATION_TYPE.value_counts() < 500].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [19]:
# Look at CLASSIFICATION value counts for binning

application_df.CLASSIFICATION.value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [20]:
# You may find it helpful to look at CLASSIFICATION value counts >1

application_df.CLASSIFICATION.value_counts()[application_df.CLASSIFICATION.value_counts() > 1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [21]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = list(application_df.CLASSIFICATION.value_counts()[application_df.CLASSIFICATION.value_counts() < 100].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64

In [9]:
# Look at NAME value counts for binning
application_df['NAME'].value_counts()

PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: NAME, Length: 19568, dtype: int64

In [22]:
# Choose a cutoff value and create a list of names to be replaced
names_to_replace = list(application_df['NAME'].value_counts()[application_df['NAME'].value_counts() < 50].index)

# Replace in dataframe
for name in names_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(
        name, "Other")

# Check to make sure binning was successful
application_df['NAME'].value_counts()

Other                                                                 24545
PARENT BOOSTER USA INC                                                 1260
TOPS CLUB INC                                                           765
UNITED STATES BOWLING CONGRESS INC                                      700
WASHINGTON STATE UNIVERSITY                                             492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                         408
PTA TEXAS CONGRESS                                                      368
SOROPTIMIST INTERNATIONAL OF THE AMERICAS INC                           331
ALPHA PHI SIGMA                                                         313
TOASTMASTERS INTERNATIONAL                                              293
MOST WORSHIPFUL STRINGER FREE AND ACCEPTED MASONS                       287
LITTLE LEAGUE BASEBALL INC                                              277
INTERNATIONAL ASSOCIATION OF LIONS CLUBS                                266
MOMS CLUB   

In [23]:
# Convert categorical data to numeric with `pd.get_dummies`

application_df = pd.get_dummies(application_df,dtype=float)
application_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  NAME_AIR FORCE ASSOCIATION  \
0       1     5000              1                         0.0   
1       1   108590              1                         0.0   
2       1     5000              0                         0.0   
3       1     6692              1                         0.0   
4       1   142590              1                         0.0   

   NAME_ALPHA PHI SIGMA  NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES  \
0                   0.0                                               0.0   
1                   0.0                                               0.0   
2                   0.0                                               0.0   
3                   0.0                                               0.0   
4                   0.0                                               0.0   

   NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES INC  \
0                                                0.0      
1                                                0.0      
2                                                0.0      
3                                                0.0      
4                                                0.0      

   NAME_AMERICAN ASSOCIATION OF UNIVERSITY WOMEN  \
0                                            0.0   
1                                            0.0   
2                                            0.0   
3                                            0.0   
4                                            0.0   

   NAME_CALIFORNIA GARDEN CLUBS INC  NAME_CISD EDUCATIONAL SUPPORT GROUPS INC  \
0                               0.0                                       0.0   
1                               0.0                                       0.0   
2                               0.0                                       0.0   
3                               0.0                                       0.0   
4                               0.0                                       0.0   

   ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0  ...                0.0                     0.0                       0.0   
1  ...                1.0                     0.0                       0.0   
2  ...                0.0                     0.0                       0.0   
3  ...                0.0                     1.0                       0.0   
4  ...                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 102 columns]

In [24]:
# Split our preprocessed data into our features and target arrays
X = application_df.drop('IS_SUCCESSFUL', axis=1).values
y = application_df['IS_SUCCESSFUL'].values

# Split the preprocessed data into a training and testing dataset
x_train,x_test, y_train, y_test = train_test_split(X, y, random_state = 123)

In [25]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
x_scaler = scaler.fit(x_train)

# Scale the data
x_train_scaled = x_scaler.transform(x_train)
x_test_scaled  = x_scaler.transform(x_test)

## Compile, Train and Evaluate the Model

In [26]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

input_features = x_train_scaled.shape[1]
hidden_nodes1  = 30
hidden_nodes2  = 60
optimized_model_1 = tf.keras.models.Sequential()

# First hidden layer
optimized_model_1.add(tf.keras.layers.Dense(units=hidden_nodes1, input_dim=input_features, activation='relu'))
# Second hidden layer
optimized_model_1.add(tf.keras.layers.Dense(units=hidden_nodes2, activation='relu'))

# Output layer
optimized_model_1.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
optimized_model_1.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 30)                3060      
                                                                 
 dense_1 (Dense)             (None, 60)                1860      
                                                                 
 dense_2 (Dense)             (None, 1)                 61        
                                                                 
Total params: 4,981
Trainable params: 4,981
Non-trainable params: 0
_________________________________________________________________


In [27]:
# Compile the model

optimized_model_1.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics=['accuracy'])

In [28]:
# Train the model

fit_model_1 = optimized_model_1.fit(x_train_scaled, y_train, validation_split = 0.15, epochs = 50)

Epoch 1/50
684/684 [==============================] - 3s 3ms/step - loss: 0.5118 - accuracy: 0.7422 - val_loss: 0.4868 - val_accuracy: 0.7587
Epoch 2/50
684/684 [==============================] - 1s 2ms/step - loss: 0.4810 - accuracy: 0.7607 - val_loss: 0.4897 - val_accuracy: 0.7600
Epoch 3/50
684/684 [==============================] - 1s 2ms/step - loss: 0.4775 - accuracy: 0.7624 - val_loss: 0.4802 - val_accuracy: 0.7665
Epoch 4/50
684/684 [==============================] - 1s 2ms/step - loss: 0.4748 - accuracy: 0.7651 - val_loss: 0.4806 - val_accuracy: 0.7634
Epoch 5/50
684/684 [==============================] - 2s 2ms/step - loss: 0.4738 - accuracy: 0.7647 - val_loss: 0.4802 - val_accuracy: 0.7670
Epoch 6/50
684/684 [==============================] - 2s 3ms/step - loss: 0.4726 - accuracy: 0.7651 - val_loss: 0.4775 - val_accuracy: 0.7668
Epoch 7/50
684/684 [==============================] - 2s 2ms/step - loss: 0.4710 - accuracy: 0.7676 - val_loss: 0.4831 - val_accuracy: 0.7673
Epoch 

In [31]:
# Evaluate the model using the test data
model_loss_1, model_accuracy_1 = optimized_model_1.evaluate(x_test_scaled, y_test, verbose = 2)
print(f"Loss: {round(model_loss_1, 3)}, Accuracy: {round(model_accuracy_1 * 100, 2)}%")

268/268 - 0s - loss: 0.4868 - accuracy: 0.7657 - 345ms/epoch - 1ms/step
Loss: 0.487, Accuracy: 76.57%


In [33]:
# Export our model to HDF5 file

optimized_model_1.save('AlphabetSoupCharity_Optimization_1.h5')

## Compile, Train and Evaluate the Model

### Model 2

In [34]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

input_features = x_train_scaled.shape[1]
hidden_nodes1  = 50
hidden_nodes2  = 100
hidden_nodes3  = 200
optimized_model_2 = tf.keras.models.Sequential()

# First hidden layer
optimized_model_2.add(tf.keras.layers.Dense(units=hidden_nodes1, input_dim=input_features, activation='relu'))
# Second hidden layer
optimized_model_2.add(tf.keras.layers.Dense(units=hidden_nodes2, activation='relu'))
# Third hidden layer
optimized_model_2.add(tf.keras.layers.Dense(units=hidden_nodes3, activation='relu'))
# Output layer
optimized_model_2.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
optimized_model_2.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 50)                5100      
                                                                 
 dense_4 (Dense)             (None, 100)               5100      
                                                                 
 dense_5 (Dense)             (None, 200)               20200     
                                                                 
 dense_6 (Dense)             (None, 1)                 201       
                                                                 
Total params: 30,601
Trainable params: 30,601
Non-trainable params: 0
_________________________________________________________________


In [35]:
# Compile the model

optimized_model_2.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics=['accuracy'])

In [36]:
# Train the model

fit_model_2 = optimized_model_2.fit(x_train_scaled, y_train, validation_split = 0.15, epochs = 75)

Epoch 1/75
684/684 [==============================] - 4s 3ms/step - loss: 0.5033 - accuracy: 0.7492 - val_loss: 0.4840 - val_accuracy: 0.7593
Epoch 2/75
684/684 [==============================] - 2s 3ms/step - loss: 0.4804 - accuracy: 0.7615 - val_loss: 0.4810 - val_accuracy: 0.7699
Epoch 3/75
684/684 [==============================] - 3s 4ms/step - loss: 0.4759 - accuracy: 0.7632 - val_loss: 0.4770 - val_accuracy: 0.7691
Epoch 4/75
684/684 [==============================] - 4s 6ms/step - loss: 0.4737 - accuracy: 0.7667 - val_loss: 0.4798 - val_accuracy: 0.7694
Epoch 5/75
684/684 [==============================] - 2s 3ms/step - loss: 0.4723 - accuracy: 0.7685 - val_loss: 0.4778 - val_accuracy: 0.7694
Epoch 6/75
684/684 [==============================] - 2s 3ms/step - loss: 0.4710 - accuracy: 0.7678 - val_loss: 0.4792 - val_accuracy: 0.7665
Epoch 7/75
684/684 [==============================] - 3s 5ms/step - loss: 0.4697 - accuracy: 0.7696 - val_loss: 0.4802 - val_accuracy: 0.7712
Epoch 

In [37]:
# Evaluate the model using the test data
model_loss_2, model_accuracy_2 = optimized_model_2.evaluate(x_test_scaled, y_test, verbose = 2)
print(f"Loss: {round(model_loss_2, 3)}, Accuracy: {round(model_accuracy_2 * 100, 2)}%")

268/268 - 0s - loss: 0.5235 - accuracy: 0.7669 - 367ms/epoch - 1ms/step
Loss: 0.524, Accuracy: 76.69%


In [38]:
# Export our model to HDF5 file

optimized_model_2.save('AlphabetSoupCharity_Optimization_2.h5')

# Analysis Report

## 1.  Overview:

A nonprofit foundation named `Alphabet Soup` aims to build a deep learning model that can help them to predict whether or not the applicants will be successful for funding. As there are only two categories i.e. successful and not successful, which makes it a binary classification problem. With the dataset provided, the goal is to build a binary classifier that could predict whether applicants will be successful if funded by Alphabet Soup.

## 2. Results

* ### Data Preprocessing:
  * What variable(s) are the target(s) for your model?
    * The *`IS_SUCCESSFUL`* column from application_df is the target variable, this is what we are trying to predict. This shows if the money was used effectively.

  * What variable(s) are the features for your model?
    * **The feature variables we used are:**
      1. `APPLICATION_TYPE`       --> Alphabet Soup application type

      2. `AFFILIATION`            --> Affiliated sector of industry

      3. `CLASSIFICATION`         --> Government organization classification

      4. `USE_CASE`               --> Use case for funding

      5. `ORGANIZATION`           --> Organization type

      6. `STATUS`                 --> Active status

      7. `INCOME_AMT`             --> Income classification

      8. `SPECIAL_CONSIDERATIONS` --> Special considerations for application

      9. `ASK_AMT`                --> Funding amount requested**

  * What variable(s) should be removed from the input data because they are neither targets nor features?
    * The *`EIN`* and *`NAME`* columns are identification columns that typically provide unique identifiers for each organization. These columns usually have no direct impact on the target variable and can be dropped without affecting the model's accuracy.

* ### Compiling, Training, and Evaluating the Model:

  * How many neurons, layers, and activation functions did you select for your neural network model, and why?
    * For this challenge I have implemented `3` different deep learning models, one of them is a simple model while the other two are the optimized versions of deep learning models with different number of hidden layers, different number on neurons, different activation functions, and different number of epochs.
        ### Simple Neural Network:  
        * The first neural network model named *`simple_model`*, is a two-layer architecture with a specific choice for the number of neurons, layers, and activation functions. By selecting 15 neurons in the first hidden layer (units_1 = 15), 25 neurons in the second hidden layer (units_2 = 25), and using the `ReLU` activation function (activation = "relu") for both hidden layers. The number of epoch for the *`simple_model`* were set 30. The choice of `ReLU` activation helps introduce non-linearity and allows the model to learn complex relationships between the input features and the target variable.

        <img src="images/simple_model.PNG" alt="Alternative text" />
        
        I used a single neuron in the output layer (units = 1) with a `sigmoid` activation function (activation = "sigmoid") to model the binary classification problem. The sigmoid activation function maps the output to a range between `0` and `1`, representing the probability of the positive class.
        
        <img src="images/simple_model_result.PNG" alt="Alternative text" />

        ### Optimized Neural Networks:
     
        * The first neural network model named *`first_model`*, is a two-layer architecture with a specific choice for the number of neurons, layers, and activation functions. By selecting 20 neurons in the first hidden layer (units_1 = 20), 30 neurons in the second hidden layer (units_2 = 30), and using the ReLU activation function (activation="relu") for both hidden layers. The number of epoch for the *`first_model`* were set 50. The choice of `ReLU` activation helps introduce non-linearity and allows the model to learn complex relationships between the input features and the target variable.

        <img src="images/first_model.PNG" alt="Alternative text" />
        
        I used a single neuron in the output layer (units = 1) with a `sigmoid` activation function (activation = "sigmoid") to model the binary classification problem. The sigmoid activation function maps the output to a range between `0` and `1`, representing the probability of the positive class.
        
        <img src="images/first_model_result.PNG" alt="Alternative text" />

        * The second neural network model named *`second_model`*, is a three-layer architecture with a specific choice for the number of neurons, layers, and activation functions. By selecting 50 neurons in the first hidden layer (units_1 = 50), 100 neurons in the second hidden layer (units_2 = 100), 200 neurons in the second hidden layer (units_2 = 200), and using the `ReLU` activation function (activation = "relu") for all the hidden layers. The number of epoch for the *`second_model`* were set 75. The choice of ReLU activation helps introduce non-linearity and allows the model to learn complex relationships between the input features and the target variable.

        <img src="images/second_model.PNG" alt="Alternative text" />
        
        I used a single neuron in the output layer (units = 1) with a `sigmoid` activation function (activation = "sigmoid") to model the binary classification problem. The sigmoid activation function maps the output to a range between `0` and `1`, representing the probability of the positive class.
        
        <img src="images/second_model_result.PNG" alt="Alternative text" />
    
  * Were you able to achieve the target model performance?
    
     * As you can see in the third I was only able to achieve 73%, which was not the target model performance which was 75%.

  * What steps did you take in your attempts to increase model performance?
    1.  **Increasing the number of neurons and epochs:**
      * By increasing the number of neurons in a layer, the model becomes more expressive and can capture complex patterns in the data. This allows for better representation of the underlying relationships between the features and the target variable, potentially leading to higher accuracy.
      
      * Increasing the number of epochs gives the model more opportunities to learn from the data and adjust the weights. It allows the model to refine its predictions and find better parameter values, which can lead to improved accuracy. However, it's important to find a balance as increasing epochs excessively can lead to overfitting.
    
    2. Removing only `EIN` and not `NAME` column:
      * In the first neural network i.e. *`simple_model`*, I removed both the `EIN` and `NAME` columns as instructed, which resulted in 73% accuracy, but in the optimization notebook, I only removed the `EIN` column, which helped in improving the accuracy from `73%` to `76%`.
      
     <img src='images/dropping_EIN.PNG' alt="Alternative text" />
     
     <img src='images/NAME_value_counts.PNG' alt="Alternative text" />


## 3. Summary:

The model built at the start, without the `EIN` and `NAME` columns, did not perform well and only gave an accuracy of `73%`. In an attempt to optimize the model, or increase the accuracy of the model, when I kept the `NAME` column and also added some extra hidden layers along with increased the number of neurons, I was finally able to acheive an accuracy greater than `75%`.

---